### ISSUE TITLE
CoreDNS 1.10.1 Return SERVFAIL Error with "dns: overflow unpacking uint16" when Upstream DNS response is more than 512 bytes

### ISSUE THREAD

**Issue by robbiezhang** [miekg/coredns#5998]
**What happened**:
Use CoreDNS 1.10.1 to query DNS `cluster-us-east-1-152269.prevops.com` using AzureDNS as upstream DNS server in an AKS cluster. It returns SERVFAIL after ~10 seconds
```
nslookup cluster-us-east-1-152269.prevops.com.
Server:         10.0.0.10
Address:        10.0.0.10#53

** server can't find cluster-us-east-1-152269.prevops.com: SERVFAIL
```

**What you expected to happen**:
It should return a list of IPs

**How to reproduce it (as minimally and precisely as possible)**:
Create an AKS cluster
Change the coredns version to 1.10.1.
Deploy a dnsutils pod, and run `nslookup cluster-us-east-1-152269.prevops.com.`

**Anything else we need to know?**:
The issue doesn't repro in 1.10.0, 1.9.4, 1.9.3 

**Environment**:
AKS, AzureDNS

- the version of CoreDNS: 1.10.1
- Corefile:
```
    .:53 {
        errors
        ready
        health {
          lameduck 5s
        }
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
          ttl 30
        }
        prometheus :9153
        forward . /etc/resolv.conf
        cache 30
        loop
        reload
        loadbalance
    }
```
- logs, if applicable:
1.10.1 (Failed)
```
[INFO] 10.244.1.4:53490 - 19745 "A IN cluster-us-east-1-152269.prevops.com. udp 54 false 512" - - 0 5.001039164s
[ERROR] plugin/errors: 2 cluster-us-east-1-152269.prevops.com. A: dns: overflow unpacking uint16
[INFO] 10.244.1.4:53490 - 19745 "A IN cluster-us-east-1-152269.prevops.com. udp 54 false 512" - - 0 5.001601765s
[ERROR] plugin/errors: 2 cluster-us-east-1-152269.prevops.com. A: dns: overflow unpacking uint16
```
1.10.0 (Succeeded)
```
[INFO] 10.244.1.4:39792 - 57873 "A IN cluster-us-east-1-152269.prevops.com. udp 54 false 512" NOERROR qr,aa,tc,rd,ra 2134 0.0001473s
[INFO] 10.244.1.4:39792 - 57873 "A IN cluster-us-east-1-152269.prevops.com. udp 54 false 512" NOERROR qr,aa,tc,rd,ra 2134 0.0001633s
[INFO] 10.244.1.4:45377 - 33354 "A IN cluster-us-east-1-152269.prevops.com. tcp 54 false 65535" NOERROR qr,aa,rd,ra 2134 0.0001247s
[INFO] 10.244.1.4:58169 - 43709 "AAAA IN cluster-us-east-1-152269.prevops.com. tcp 54 false 65535" NOERROR qr,aa,rd,ra 149 0.0000537s
```
- OS (e.g: `cat /etc/os-release`):
```
PRETTY_NAME="Ubuntu 22.04.2 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.2 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```
- Others:
I can provide tcpdump if needed for debugging.

**Comment by robbiezhang** [comment]
Note that setting the bufsize to a higher values (e.g. 1232) doesn't help in this case.

**Comment by chrisohaver** [comment]
Just to test (I am not suggesting this as a fix), does disabling _cache_ make the issue go away? (#5671)

**Comment by phealy** [comment]
@chrisohaver A few more notes on this one (I work with @robbiezhang):

1. This only happens with clients that aren't using eDNS0 - `dig cluster-us-east-1-152269.prevops.com.` works, but `dig +noedns cluster-us-east-1-152269.prevops.com.` doesn't.
1. The DNS server in this case is not complying with standards - there's no OPT RR in the request (so no eDNS0), but the response is coming back in a single ~700 byte UDP packet instead of being truncated appropriately. This is preventing CoreDNS from handling the request in the 512-byte buffer.

This means that #5671 actually is the problem - but that's because before that change, the cache plugin being enabled would *mask* the result of a non-compliant upstream server by forcing a 2k buffer, even if we weren't using eDNS0.

[code from plugin/cache/handler.go that was removed in #5671/v1.10.1]:(https://github.com/coredns/coredns/blob/v1.10.0/plugin/cache/handler.go#L173-L175):
```go
// defaultUDPBufsize is the bufsize the cache plugin uses on outgoing requests that don't
// have an OPT RR.
const defaultUDPBufSize = 2048
```

I confirmed this by rolling back to 1.9.4 with a minimal configuration - it had the same problem. I added the cache plugin and suddenly the query worked appropriately.

In other words, this is related to, or potentially even a duplicate of, #5953.

**Comment by phealy** [comment]
To be clear - prior to #5671, the cache plugin would use a larger buffer, get the response from the misbehaving upstream appropriately, then truncate it and send it back to the client. The client would then make the same request over TCP, which would get answered from the cache with the full record.

**Comment by chrisohaver** [comment]
Thanks that makes sense. I don't think this was an intentional feature of the prior mechanism (which was problematic in several other ways).  IIUC, before the prior mechanism was introduced (in https://github.com/coredns/coredns/pull/4085), the same thing would have happened.  

IOW, I don't think there is a problem with #5671.

> this is related to, or potentially even a duplicate of, https://github.com/coredns/coredns/issues/5953.

It's definitely related... #5953 is identifying the underlying cause of non-compliant upstreams and is proposing a fix. 

Although in #5953, the issue is observed in 1.10.0 ... and here the issue is not observed in 1.10.0.

**Comment by phealy** [comment]
To clarify - I didn't mean that I thought #5671 was broken, I just meant that it was partially covering up the problem before and now it's not.

I suspect the problem that #5953 is having, reading the linked Red Hat bug, is that the UDP response being returned (and breaking with 1.10.0) is >2048 bytes. The server I'm working with does return truncate at ~1024 bytes even without eDNS0, so I can't reproduce that failure case.

Basically:
post #4085, 1.10.0 and earlier: cache plugin forces the buffer size to 2048 by default. Non-eDNS0 clients that hit a noncompliant upstream will get an error if the UDP response is >2048 bytes. If the response is >512 but <2048 bytes, coredns will properly return the response by truncating and then dealing with cache. This is the case we fell into, because we'd only get non-compliant responses if the response was >512 but <1024 bytes.

1.10.1+: cache plugin now does not force the buffer size to 2048, so the packet size for hitting the problem with a noncompliant DNS server is now >512 bytes over UDP.

**Comment by chrisohaver** [comment]
I think we're on the same page.

The fix proposed in https://github.com/coredns/coredns/issues/5953 would also fix the issue reported in this issue.

**Comment by SriHarsha001** [comment]
@chrisohaver  

The code fix for this issue is merged in this PR - https://github.com/coredns/coredns/pull/6277

I wanted to check with you on when do we expect the release of next CoreDNS version. 

Any info will be greatly appreciated.

**Comment by ronyrv13** [comment]
Hello @SriHarsha001 / @chrisohaver,

I have been using CoreDNS version 1.10.1 for the past six months. However, recently, I received reports of DNS failures from other teams. 
Upon examining the CoreDNS logs, I observed intermittent failures, specifically with a **SERVFAIL** status.

> Our setup involves having a CoreDNS forwarder in each region. These forwarders direct DNS requests to an upstream CoreDNS HUB server. 
> For internet-bound DNS queries, the requests are further forwarded to either Google or Quad9 DNS servers.

Upon a thorough review of the CoreDNS logs, it became apparent that the majority (99%) of SERVFAIL entries were associated with **TCP** DNS requests having a **bufsize of 65535**.

_Log Sample:_ 
```\"A IN example.com tcp 77 true 65535\" SERVFAIL qr,aa,rd,ra 66 0.000119001s\n","stream":"stdout"```

Is this particular problem related to the above subject issue (#5998) ?
_Also, I have no idea why they are TCP DNS requests while the client must have performed UDP (default) DNS Queries._
If yes, then do we have any temp solution to this like setting ``bufsize 4096`` in CoreFile.

Any insights you can provide on this matter would be greatly appreciated.

**Comment by ronyrv13** [comment]
I am observing the previously mentioned behavior only for specific domains. 
As previously stated, we have a **CoreDNS** forwarder in each region, these forwarders direct DNS requests to an upstream **CoreDNS** HUB server. 

Upon inspecting the logs on the forwarder, I observed a UDP DNS request. 
However, when I examined the corresponding DNS request log on the CoreDNS HUB server, it appeared as a TCP request with a bufsize of 65535.

@chrisohaver, I know you are busy, but any insights you can provide on this matter would be greatly appreciated.
